drivers/ieee802154/ieee802154_efr32.c

```diff
@@ -0,0 +1,558 @@
+    }
+    else if (type == IEEE802154_FILTER_TYPE_PAN_ID)
+    {
+        return efr32_set_pan_id(dev, filter->pan_id);
+    }
+
+    return -ENOTSUP;
+}
+
+static int efr32_set_txpower(struct device *dev, s16_t dbm)
+{
+    struct efr32_context *efr32 = dev->driver_data;
+
+    RAIL_Status_t status;
+    status = RAIL_SetTxPower(efr32->rail_handle, dbm);
+
+    if (status != RAIL_STATUS_NO_ERROR)
+    {
+        return -EIO;
+    }
+
+    return 0;
+}
+
+static int efr32_start(struct device *dev)
+{
+    struct efr32_context *efr32 = dev->driver_data;
+    ARG_UNUSED(efr32);
+
+    efr_state = EFR32_STATE_SLEEP;
+
+    LOG_DBG("EFR32 802154 radio started");
+
+    return 0;
+}
+
+static int efr32_stop(struct device *dev)
+{
+    struct efr32_context *efr32 = dev->driver_data;
+    ARG_UNUSED(efr32);
+
+    efr_state = EFR32_STATE_DISABLED;
+
+    LOG_DBG("EFR32 802154 radio stopped");
+
+    return 0;
+}
+
+static int efr32_tx(struct device *dev, struct net_pkt *pkt,
+                    struct net_buf *frag)
+{
+    struct efr32_context *efr32 = dev->driver_data;
+
+    u8_t payload_len = net_pkt_ll_reserve(pkt) + frag->len;
+    u8_t *payload = frag->data - net_pkt_ll_reserve(pkt);
+
+    RAIL_TxOptions_t tx_opts = RAIL_TX_OPTIONS_NONE;
+    RAIL_Status_t status;
+
+    LOG_DBG("%p (%u)", payload, payload_len);
+
+    efr32->tx_success = false;
+
+    memcpy(efr32->tx_buf + 1, payload, payload_len);
+    efr32->tx_buf[0] = payload_len + EFR32_FCS_LENGTH;
+
+    /* Reset semaphore in case ACK was received after timeout */
+    k_sem_reset(&efr32->tx_wait);
+
+    RAIL_WriteTxFifo(efr32->rail_handle, efr32->tx_buf, payload_len, true);
+    status = RAIL_StartTx(efr32->rail_handle, 0, tx_opts, NULL);
+
+    if (status != RAIL_STATUS_NO_ERROR)
+    {
+        LOG_ERR("Cannot start tx");
+        return -EIO;
+    }
+
+    LOG_DBG("Sending frame");
+
+    if (k_sem_take(&efr32->tx_wait, ACK_TIMEOUT))
+    {
+        LOG_DBG("ACK not received");
+        return -EIO;
+    }
+
+    LOG_DBG("Result: %d", efr32->tx_success);
+
+    return efr32->tx_success ? 0 : -EBUSY;
+}
+
+/* 
+    Continously runs and tries to fetch packets..
+*/
+static void efr32_rx(int arg)
+{
+    struct device *dev = INT_TO_POINTER(arg);
+    struct efr32_context *efr32 = dev->driver_data;
+
+    while (1)
+    {
+        LOG_DBG("Waiting for RX free");
+        k_sem_take(&efr32->rx_wait, K_FOREVER);
+
+        RAIL_Idle(efr32->rail_handle, RAIL_IDLE_ABORT, true);
+        RAIL_StartRx(efr32->rail_handle, efr32->channel, NULL);
+    }
+}
+
+static int efr32_init(struct device *dev)
+{
+    struct efr32_context *efr32 = dev->driver_data;
+    RAIL_Handle_t rail_handle;
+    RAIL_Status_t status;
+    rail_handle = RAIL_Init(&s_rail_config, NULL);
+
+    k_sem_init(&efr32->rx_wait, 0, 1);
+    k_sem_init(&efr32->tx_wait, 0, 1);
+
+    if (rail_handle == NULL)
+    {
+        LOG_ERR("Unable to init");
+        return -EIO;
+    }
+
+    RAIL_DataConfig_t rail_data_config = {
```

Comment:
Always declare a variable at the beginning of its code block, apply that everywhere